mono/utils/mono-merp.c

```diff
@@ -65,21 +65,40 @@ kernel_version_string (void)
 }
 
 static gboolean
-starts_with(const char *pre, const char *str)
+starts_with (const char *pre, const char *str)
 {
-    return strncmp(pre, str, strlen(pre)) == 0;
+    return strncmp (pre, str, strlen (pre)) == 0;
 }
 
 static const char *
 macos_version_string (void)
 {
+	/* Can't allocate in here, could be called from a signal handler in a
+	 * crashed process.
+	 */
 	static const char *version_string;
+	static char buf[256];
+	static const size_t buf_size = sizeof (buf);
+
+	if (version_string)
+		return version_string;
+
 	/* macOS 10.13.6 or later */
 	if (!version_string) {
-		char s[256];
-		size_t size = sizeof (s);
-		if (sysctlbyname ("kern.osproductversion", s, &size, NULL, 0) >= 0)
-			version_string = strdup(s);
+		size_t size = 0;
+		if (sysctlbyname ("kern.osproductversion", NULL, &size, NULL, 0) < 0) {
+			buf[0] = '\0';
```

Comment:
Should this be buf[size]?

Reply:
Nope nevermind, this is the error case.

Reply:
No, this is the error path.  If we can't even get the expected size of the product version value, set buf to `""` and return that.